ui/src/app/edge/live/common/grid/modal/modal.ts

```diff
@@ -1,53 +1,146 @@
-import { ChannelAddress, CurrentData, EdgeConfig, GridMode } from 'src/app/shared/shared';
+import { TextIndentation } from 'src/app/shared/genericComponents/modal/modal-line/modal-line';
+import { ChannelAddress, Edge, EdgeConfig, GridMode, Utils } from 'src/app/shared/shared';
+import { Role } from 'src/app/shared/type/role';
+
+export type FormlyFieldLine = {
+  type: string,
+  channel?: string,
+  filter?: Function,
+  converter?: Function,
+  name?: string,
+  nameSuffix?: Function,
+  indentation?: TextIndentation,
+  roleIsAtLeast?: Role,
+  children?: FormlyFieldLine[]
+}
 
 @Component({
   selector: 'modal',
   templateUrl: './modal.html'
 })
 export class ModalComponent extends AbstractModal {
 
-  protected readonly GridMode = GridMode;
-  protected grid: { mode: GridMode, buyFromGrid: number, sellToGrid: number } = { mode: GridMode.UNDEFINED, buyFromGrid: null, sellToGrid: null };
-
-  protected meters: { component: EdgeConfig.Component, isAsymmetric: boolean }[] = []
+  protected form: FormGroup = new FormGroup({});
+  protected fields: FormlyFieldConfig[] = [];
 
   protected override getChannelAddresses(): ChannelAddress[] {
-    let channelAddresses: ChannelAddress[] = [];
-
-    const asymmetricMeters = this.config.getComponentsImplementingNature("io.openems.edge.meter.api.AsymmetricMeter")
-      .filter(comp => comp.isEnabled && this.config.isTypeGrid(comp))
-
-    this.config.getComponentsImplementingNature("io.openems.edge.meter.api.SymmetricMeter")
-      .filter(component => component.isEnabled && this.config.isTypeGrid(component))
-      .forEach(component => {
-        var isAsymmetric = asymmetricMeters.filter(element => component.id == element.id).length > 0;
-        this.meters.push({ component: component, isAsymmetric: isAsymmetric });
-      })
-
-    channelAddresses.push(
-      new ChannelAddress("_sum", 'GridMode'),
-      new ChannelAddress('_sum', 'GridActivePower'),
-      new ChannelAddress('_sum', 'GridActivePowerL1'),
-      new ChannelAddress('_sum', 'GridActivePowerL2'),
-      new ChannelAddress('_sum', 'GridActivePowerL3'),
-    )
-    return channelAddresses;
+    this.edge.getConfig(this.websocket).pipe(filter(config => !!config)).subscribe((config) => {
+      this.fields = ModalComponent.generateGridModal(this.edge.id, config, this.edge.role, this.translate);
+    })
+    return [];
   }
 
-  protected override onCurrentData(currentData: CurrentData): void {
+  public static generateModalMeterPhases(component: EdgeConfig.Component, translate: TranslateService, userRole: Role): FormlyFieldLine[] {
+    let fields: FormlyFieldLine[] = [];
 
-    this.grid.mode = currentData.allComponents["_sum/GridMode"]
-    let gridActivePower = currentData.allComponents['_sum/GridActivePower']
-    this.grid.buyFromGrid = gridActivePower > 0 ? gridActivePower : 0;
-    this.grid.sellToGrid = gridActivePower < 0 ? (gridActivePower * -1) : 0;
+    for (let phase of ['L1', 'L2', 'L3']) {
```

Comment:
I tend to prefer `['L1', 'L2', 'L3'].forEach(phase => {`

Reply:
why do you prefer that?